### ISSUE TITLE
Library dependencies ordering causing inconsistent assumptions over interface

### ISSUE THREAD

**Issue by horlogeislux** [ocaml/ocaml#13218]
Hello,

This is a follow-up from [an issue](https://github.com/ocaml/dune/issues/10606#issuecomment-2148077758) I reported to `dune`, where a contributor recommended I should report it here.

I am using `dune` as my build system for an OCaml project (the `Zelus` compiler).
I recently encountered a bug where `ocamlopt` silently links my executable with the wrong interface when given a specific ordering of libraries (unexpected and undesired), but prints an error with the other ordering (expected and desired).
Thank you for your guidance.

## Expected Behavior

I would expect a "Files ... make inconsistent assumptions over interface" error in the minimal working example below, regardless of the order of `lib_a` or `lib_b`.

## Actual Behavior

The behavior I faced was that in one ordering (`lib_a` `lib_b`), `ocamlopt` prints out the helpful error message ("Files [...] make inconsistent assumptions over interface"), but with another ordering (`lib_b` `lib_a`), `ocamlopt` lets `lib_a`'s code prevail silently.

- Should this be caught by `ocamlopt` every time or is this behavior justified?

Moreover, if I give `ocamlopt` the `-linkall` flag, I get a more descriptive error message:
`Error: Files lib_b/lib_b.cmxa and lib_a/lib_a.cmxa both define a module named Common.`

- Should this be the default behavior for `ocamlopt` ?

## Reproduction

1. `git clone` the repository below, make sure you are on branch `wrap_false` (the default branch)
2. `make`: outputs an error at link time (expected and desired)
3. In ./dune, uncomment the second version of the `libraries`, in the `executable` stanza
4. `make`: does not output an error, and uses `lib_a`'s code (unexepected and undesired)

With `-linkall` flag in `dune`:
1. `git switch linkall`
2. `make`: outputs a more helpful error at link time, which I think is clearer and thus preferable

## Specifications

- Version of `dune` (output of `dune --version`): 3.15.3
- Version of `ocaml` (output of `ocamlc --version`): 4.14.2
- Operating system (distribution and version): Ubuntu 22.04.4 LTS; 6.5.0-35-generic; x86_64

## Additional information 

- Link to gist with verbose output (run `dune` with the `--verbose` flag): [the two dune build gists](https://gist.github.com/horlogeislux/2e3b63b450463b4ccd35bf3783031844)
- The minimal working example [https://github.com/horlogeislux/liborder-dune](https://github.com/horlogeislux/liborder-dune)

**Comment by lthls** [comment]
I suspect this is a variation of #10564, which was fixed in #11635. Could you check if the problem still exists on OCaml 5.2 ?

**Comment by horlogeislux** [comment]
Yes the problem still exists on OCaml 5.2.
I added a new terminal output on the [gist](https://gist.github.com/horlogeislux/2e3b63b450463b4ccd35bf3783031844#file-liborder-b-a-ocaml-5-2), that shows the output of
`lsb_release -a && ocamlopt --version && dune --version && dune clean && dune build --verbose ; cat bin/dune`

(edit 14h01: I enabled the correct line in `bin/dune` and changed the gist to account for this)

**Comment by gasche** [comment]
My impression from a distance is that this may be the desired behavior -- but it is hard to tell without more details / looking at the code. If no one is explicitly depending on `Lib_b.Common`, then it does not get linked into the final executable and its code does not run.

Deep down the problem is that you are using a build system that guesses the dependencies, instead of being told about them, without having enough information in the source to make the right decisions. Some guesses result in incompatible assumptions, so you get an error, some guesses result in something that makes sense but does not do what you expect. You need to be more explicit in the source or to use a build system that gives you more control.

Note: using Dune's `wrapped` option  (which is the default) avoids the problem in two different ways: (1) giving different names to the compilation units, so that both `Common` modules can be used simultaneously, and (2) forcing you in the source to say `Lib_a.Common` or `Lib_b.Common`, so that you are explicit about which you mean. It is a reasonable way to avoid this issue -- there are others (basically doing something similar by hand with module aliases), but they are more work and dune does not support them well.

**Comment by gasche** [comment]
As to whether this is expected -- you ask whether different defaults would give a better user experience, which is a good question. Yes, users do rely on the fact that binaries are small because unused modules are not linked in the binary.

I think that it would be reasonable to get a warning/error when two modules of the same name are available in the search path, and some user code may ambiguously refer to one or the other without a clear way to tell. The problem is to define a notion of precedence/ordering between available choices in this situation, which aligns with user expectations.

For example, if they have a local module Common, and they use a library that also has a module Common, they probably expect their local Common module to be resolved in priority -- this is what happens. But if they use `-I lib1 -I lib2`, the compiler currently searches for Common in `lib1` first and `lib2` second, so `lib1` takes precedence, this may not be what the user expects (especially if the -I order is hidden by the build system) and a warning would be reasonable. (So: no ordering between independent -I invocations, but the implicit `-I .` part of the load path takes precedence.)

Then the question is: if a user is in this situation of two libraries/subdirectories exposing the same module and there is potential for confusion, how do they fix it? Currently there is no way at the source level to say "I mean compilation unit `liba/Common` rather than `libb/Common`", which we could ask the user to use to disambiguate. Currently this must reside in the build-system configuration, and Dune understandably prefers to encourages people to use `(wrapped true)` than to try to solve these tricky problems.

(What would source-level intent look like? We could have something like a toplevel directlive `#include "liba/common"` which would be a counterpart of an explicit `-cmi-path liba/common.cmi`, used as a "unit header".)

**Comment by horlogeislux** [comment]
Thank you for your clear answers.

Yes the `dune` build system guesses dependencies, sometimes [sorting the flags alphabetically](https://github.com/ocaml/dune/issues/10606#issuecomment-2149305322), sometimes following the order specified in the `dune` file. This mismatch caused an asymmetry that was hard to catch, and that the compiler unfortunately did not warn about at link time.

I would love to use `dune`'s `wrapped` option to enable this disambiguation, but I am currently contributing to an existing project which does not use it, because it was created before `dune` existed and never made fully up-to-date with current `dune` defaults. The use of `(wrapped false)` is then, I believe, the root cause of the issue.

> Then the question is: if a user is in this situation of two libraries/subdirectories exposing the same module and there is potential for confusion, how do they fix it?

The fix I found currently is by specifying the correct library last, for instance `(libraries lib_to_be_shadowed lib_that_has_the_correct_interface)`.
However, this fix is not pretty, especially if `lib_a` and `lib_b` contain multiple modules, and if we want `lib_a/common.ml` to shadow `lib_b/common.ml` _and_ `lib_a/other.ml` to _be shadowed_ by `lib_b/other.ml`. We could then split both libraries in sub-libraries (if the causality of dependencies allow it), but it really does not seem adequate.

> I think that it would be reasonable to get a warning/error when two modules of the same name are available in the search path, and some user code may ambiguously refer to one or the other without a clear way to tell. The problem is to define a notion of precedence/ordering between available choices in this situation, which aligns with user expectations.

I would love for to this to exist, at least as an optional warning.

**Comment by gasche** [comment]
> The use of `(wrapped false)` is then, I believe, the root cause of the issue.

Yes. If you really want to use dune, I would recommend taking the time to restructure the project layout to be able to use `(wrapped true)` instead.

**Comment by gasche** [comment]
> I would love for to this to exist, at least as an optional warning.

The problem: if the warning is disabled by default, users never see it or learn of it, and it is not useful. If it is enabled by default, then it annoys users and they want to fix the issue. But in this case I don't know how to fix the issue. Here it is a single project that is in control of both sublibraries, but in general this may not be case. What is the recommendation to give if people use two libraries that share the same name for a compilation unit? ("Give up on one of the libraries"?) So I think that we would need a mechanism for users to disambiguate at the source level first, before we start complaining about ambiguities.

**Comment by horlogeislux** [comment]
> > I would love for to this to exist, at least as an optional warning.
> 
> The problem: if the warning is disabled by default, users never see it or learn of it, and it is not useful. If it is enabled by default, then it annoys users and they want to fix the issue. But in this case I don't know how to fix the issue. Here it is a single project that is in control of both sublibraries, but in general this may not be case. What is the recommendation to give if people use two libraries that share the same name for a compilation unit? ("Give up on one of the libraries"?) So I think that we would need a mechanism for users to disambiguate at the source level first, before we start complaining about ambiguities.

Oh, now I fully get why you evoked this priority, which must come before and not just as a "nice to have concurrently".

**Comment by gasche** [comment]
(I'm taking the liberty to close the issue as "won't fix" for now because either one considers that `(wrapped true)` is a fix, and then there is nothing to do, or we want a language/compiler-level solution, and we don't have a design proposal for what to do. This does not mean that the issue is not a real issue, but rather that it's not worth pinging people about it regularly before we come up with a more concrete plan.)
